src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java

```diff
@@ -0,0 +1,145 @@
+    // Loop through the pages.
+    // We don't know how many pages there are beforehand. We stop looping when the new page
+    // doesn't have any results.
+    int page = 0;
+    do {
+      URI uri;
+      try {
+        URIBuilder builder = new URIBuilder("https://api.umd.io/v1/courses");
+        builder.setParameter("page", Integer.toString(page++));
+        uri = builder.build();
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+      JSONArray coursesArray = apiUtil.getJsonArray(uri);
+      if (coursesArray == null) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      } else if (coursesArray.size() == 0) {
+        respondWithError(HttpServletResponse.SC_OK, response);
+        return;
+      }
+      addCourses(coursesArray, response);
+    } while (true);
+  }
+
+  /**
+   * Creates a Course object from a JSONArray.
+   * @param coursesArray The course json from the UMD API.
+   * @param response The HttpServletResponse object.
+   */
+  private void addCourses(JSONArray coursesArray, HttpServletResponse response) throws IOException {
+    if (coursesArray == null) {
+      return;
+    }
+
+    // Loop through each page's courses.
+    for (Object jsonObject : coursesArray) {
+      JSONObject courseJson = (JSONObject) jsonObject;
+
+      // Create a Course object from the JSONObject for parsing.
+      Course course = new Course(courseJson);
+
+      URI uri;
+      try {
+        uri = new URI("https://api.umd.io/v1/courses/" + course.getCourseId() + "/sections");
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+      JSONArray sectionsArray = apiUtil.getJsonArray(uri);
+      addSectionsToCourse(sectionsArray, course);
+    }
+  }
+
+  /**
+   * Adds the Course and Section to datastore.
+   * @param sectionsArray The section json from the UMD API.
+   * @param course The Course object.
+   */
+  private void addSectionsToCourse(JSONArray sectionsArray, Course course) {
+    if (sectionsArray == null || course == null) {
+      return;
+    }
+
+    // Loop through each course's sections.
+    ArrayList<EmbeddedEntity> sectionEntities = new ArrayList<EmbeddedEntity>();
+    for (Object jsonObject : sectionsArray) {
+      // TODO(savsa): create Section class to handle JSON object parsing.
+      // For now, just store the section id.
+      JSONObject sectionJson = (JSONObject) jsonObject;
+      EmbeddedEntity sectionEntity = new EmbeddedEntity();
+      sectionEntity.setProperty("section_id", (String) sectionJson.get("section_id"));
+      sectionEntities.add(sectionEntity);
+    }
+
+    Entity entity = new Entity("Course");
+    entity.setProperty("course_id", course.getCourseId());
```

Comment:
what happens if you run this function twice? It seems like you'll get duplicated entities.

Reply:
how would you prevent that? by using keys?

Reply:
Keys is one way. But more likely, you'd want to look up any existing records and update it with your newly pulled data.

Reply:
Actually I think because I'm using `datastore.put()` instead of `datastore.add()`, it overwrites the previous entry if it already existed.

Reply:
Which fields are you using to determine whether it "already existed"?

Reply:
course_id?

Reply:
It looks like datastore does "overwrite" based on the key of each object. Is `course_id` defined as the key? It doesn't look like it.

https://cloud.google.com/appengine/docs/standard/java/datastore/creating-entities#Updating_entities

Reply:
I was slightly off about put vs add. But I changed my approach anyways. if it finds that the entity already exists, then that entity gets edited with the new properties (the key stays the same), and then it gets re `put` into the database. And since I edited the entity that was already in the database and the key is the same, then it overwrites that entity with the new modified one. If the entity doesn't already exists, it creates a new one and then `put`s it into the database.

I added an extra test that tests this and the code passes.